File: VB/DataBindingToListExample/WeatherReport.vb
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel

Namespace DataBindingToListExample
    #Region "#WeatherReport"
    Public Class WeatherReport
        <DisplayName("Date")> _
        Public Property [Date]() As Date
        <DisplayName("Weather Condition")> _
        Public Property Weather() As Weather
        <DisplayName("Max and Min Temperature")> _
        Public Property HourlyReport() As List(Of HourlyReport)
    End Class

    Public Class HourlyReport
        Public Property Hour() As Integer
        Public Property Temperature() As Integer
    End Class

    Public Enum Weather
        Sunny
        Cloudy
        Windy
        Gloomy
        Foggy
        Misty
        Rainy
        Undefined
    End Enum
    #End Region ' #WeatherReport
    Public NotInheritable Class MyWeatherReportSource

        Private Sub New()
        End Sub

        Private Shared rand As Random = New System.Random()

        Private Shared data_Renamed As List(Of WeatherReport)
        Private Shared dataBindingList As BindingList(Of WeatherReport)

        Public Shared ReadOnly Property Data() As List(Of WeatherReport)
            Get
                If data_Renamed Is Nothing Then
                    data_Renamed = GetReport()
                End If
                Return data_Renamed
            End Get
        End Property
        Public Shared ReadOnly Property DataAsBindingList() As BindingList(Of WeatherReport)
            Get
                If dataBindingList Is Nothing Then
                    dataBindingList = New BindingList(Of WeatherReport)(Data)
                End If
                Return dataBindingList
            End Get
        End Property
        Public Shared Function GetReport() As List(Of WeatherReport)
            Dim report = New List(Of WeatherReport)()

            report.Add(New WeatherReport() With {.Date = Date.Today, .Weather = Weather.Rainy, .HourlyReport = GenerateRandomHourlyReport()})

            report.Add(New WeatherReport() With {.Date = Date.Today.AddDays(-1), .Weather = Weather.Cloudy, .HourlyReport = GenerateRandomHourlyReport()})

            report.Add(New WeatherReport() With {.Date = Date.Today.AddDays(-2), .Weather = Weather.Sunny, .HourlyReport = GenerateRandomHourlyReport()})

            report.Add(New WeatherReport() With {.Date = Date.Today.AddDays(-3), .Weather = Weather.Gloomy, .HourlyReport = GenerateRandomHourlyReport()})
            Return report
        End Function
        Public Shared Function GenerateRandomHourlyReport() As List(Of HourlyReport)
            Dim report = New List(Of HourlyReport)()

            For i As Integer = 0 To 23
                Dim hourlyReport = New HourlyReport()
                hourlyReport.Hour = i
                hourlyReport.Temperature = rand.Next(30)
                report.Add(hourlyReport)
            Next i
            Return report
        End Function
        Public Shared Sub Reload()
            data_Renamed = GetReport()
        End Sub
    End Class
End Namespace

File: CS/DataBindingToListExample/WeatherReport.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;

namespace DataBindingToListExample {
    #region #WeatherReport
    public class WeatherReport {
        [DisplayName("Date")]
        public DateTime Date { get; set; }
        [DisplayName("Weather Condition")]
        public Weather Weather { get; set; }
        [DisplayName("Max and Min Temperature")]
        public List<HourlyReport> HourlyReport { get; set; }
    }

    public class HourlyReport {
        public int Hour { get; set; }
        public int Temperature { get; set; }
    }

    public enum Weather {
        Sunny,
        Cloudy,
        Windy,
        Gloomy,
        Foggy,
        Misty,
        Rainy,
        Undefined
    }
    #endregion #WeatherReport
    public static class MyWeatherReportSource {
        private static Random rand = new System.Random();
        static List<WeatherReport> data;
        static BindingList<WeatherReport> dataBindingList;

        public static List<WeatherReport> Data
        {
            get {
                if (data == null) {
                    data = GetReport();
                }
                return data;
            }
        }
        public static BindingList<WeatherReport> DataAsBindingList
        {
            get {
                if (dataBindingList == null) {
                    dataBindingList = new BindingList<WeatherReport>(Data);
                }
                return dataBindingList;
            }
        }
        public static List<WeatherReport> GetReport() {
            var report = new List<WeatherReport>();

            report.Add(new WeatherReport() {
                Date = DateTime.Today,
                Weather = Weather.Rainy,
                HourlyReport = GenerateRandomHourlyReport()
            });

            report.Add(new WeatherReport() {
                Date = DateTime.Today.AddDays(-1),
                Weather = Weather.Cloudy,
                HourlyReport = GenerateRandomHourlyReport()
            });

            report.Add(new WeatherReport() {
                Date = DateTime.Today.AddDays(-2),
                Weather = Weather.Sunny,
                HourlyReport = GenerateRandomHourlyReport()
            });

            report.Add(new WeatherReport() {
                Date = DateTime.Today.AddDays(-3),
                Weather = Weather.Gloomy,
                HourlyReport = GenerateRandomHourlyReport()
            });
            return report;
        }
        public static List<HourlyReport> GenerateRandomHourlyReport() {
            var report = new List<HourlyReport>();

            for (int i = 0; i < 24; i++) {
                var hourlyReport = new HourlyReport();
                hourlyReport.Hour = i;
                hourlyReport.Temperature = rand.Next(30);
                report.Add(hourlyReport);
            }
            return report;
        }
        public static void Reload() {
            data = GetReport();
        }
    }
}
